File: Alkitab/src/main/java/yuku/alkitab/base/util/DownloadMapper.java
package yuku.alkitab.base.util;

import android.app.DownloadManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;
import android.text.format.Formatter;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import com.downloader.Error;
import com.downloader.OnDownloadListener;
import com.downloader.PRDownloader;
import com.downloader.Status;
import com.downloader.request.DownloadRequest;
import java.io.File;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;
import yuku.alkitab.base.App;
import yuku.alkitab.base.ac.AlertDialogActivity;
import yuku.alkitab.base.br.VersionDownloadCompleteReceiver;
import yuku.alkitab.debug.R;
import yuku.alkitab.versionmanager.VersionListFragment;

public enum DownloadMapper {
    instance;

    static final String TAG = DownloadMapper.class.getSimpleName();
    static final String NOTIF_TAG = "DownloadMapper";
    static final String NOTIFICATION_CHANNEL_ID = "download_mapper";

    static class Row {
        public int id;
        public String key;
        public String title;
        public String destPath;
        public long currentBytes;
        public long totalBytes;
        /**
         * When is the last time the progress is notified (in currentTimeMillis).
         * We should not notify about progress too often.
         */
        public long previouslyNotifiedTime;
        public Map<String, String> attrs;
    }

    final Map<String, Row> currentByKey = new LinkedHashMap<>();
    final Map<Integer, Row> currentById = new LinkedHashMap<>();

    public int getStatus(final String downloadKey) {
        final Row row = currentByKey.get(downloadKey);
        return getStatus(row);
    }

    public int getStatus(final int id) {
        final Row row = currentById.get(id);
        return getStatus(row);
    }

    public float getDownloadProgress(final String downloadKey) {
        final Row row = currentByKey.get(downloadKey);
        return getDownloadProgress(row);
    }

    public float getDownloadProgress(final int id) {
        final Row row = currentById.get(id);
        return getDownloadProgress(row);
    }

    static int prToDmStatus(Status status) {
        return switch (status) {
            case QUEUED -> DownloadManager.STATUS_PENDING;
            case RUNNING -> DownloadManager.STATUS_RUNNING;
            case PAUSED -> DownloadManager.STATUS_PAUSED;
            case COMPLETED -> DownloadManager.STATUS_SUCCESSFUL;
            case CANCELLED -> DownloadManager.STATUS_FAILED;
            default -> 0;
        };
    }

    private int getStatus(final Row row) {
        if (row == null) {
            return 0;
        } else {
            final Status status = PRDownloader.getStatus(row.id);
            if (status == Status.UNKNOWN || status == null) {
                // stale data found. Remove immediately
                currentByKey.remove(row.key);
                currentById.remove(row.id);
                return 0;
            } else {
                return prToDmStatus(status);
            }
        }
    }

    private float getDownloadProgress(final Row row) {
        if (row == null) {
            return -1.f;
        } else {
            return PRDownloader.getDownloadProgress(row.id);
        }
    }

    /**
     * Must be called only after verifying that this id exists.
     */
    public Map<String, String> getAttrs(final int id) {
        final Row row = currentById.get(id);
        return row.attrs;
    }

    /**
     * Must be called only after verifying that this id exists.
     */
    public String getDownloadedFilePath(final int id) {
        final Row row = currentById.get(id);
        return row.destPath;
    }

    static String downloadTempDirPath() {
        final File cacheDir = App.context.getCacheDir();
        final File res = new File(cacheDir, "DownloadMapper-tmp");
        //noinspection ResultOfMethodCallIgnored
        res.mkdirs();
        return res.getAbsolutePath();
    }

    @NonNull
    static String downloadTempBasename(final String downloadKey) {
        // Return value must be filename-safe.
        // So I will remove all non-safe characters and append the hashcode of the downloadKey.
        final StringBuilder safe = new StringBuilder(downloadKey.length() + 40);
        for (char c : downloadKey.toCharArray()) {
            if ('A' <= c && c <= 'Z' || 'a' <= c && c <= 'z' || '0' <= c && c <= '9') {
                safe.append(c);
            }
        }

        return "DownloadMapper-" + safe + downloadKey.hashCode() + ".tmp";
    }

    public void enqueue(final String downloadKey, final String url, final String title, final Map<String, String> attrs) {
        final String downloadTempDirPath = downloadTempDirPath();
        final String downloadTempBasename = downloadTempBasename(downloadKey);

        final DownloadRequest req = PRDownloader.download(url, downloadTempDirPath, downloadTempBasename).build();

        final Row row = new Row();

        req.setOnProgressListener(progress -> {
            Log.d(TAG, "@@onProgress " + progress.currentBytes + "/" + progress.totalBytes);
            row.currentBytes = progress.currentBytes;
            row.totalBytes = progress.totalBytes;

            // only notify if last notification is not recent enough.
            final long now = System.currentTimeMillis();
            if (row.previouslyNotifiedTime == 0 || now - row.previouslyNotifiedTime > 1000) {
                row.previouslyNotifiedTime = now;

                displayNotifs();
                App.getLbm().sendBroadcast(new Intent(VersionListFragment.ACTION_RELOAD));
            }
        });

        final int[] p_id = {0};
        p_id[0] = req.start(new OnDownloadListener() {
            @Override
            public void onDownloadComplete() {
                final int id = p_id[0];
                VersionDownloadCompleteReceiver.onReceive(id);
            }

            @Override
            public void onError(final Error error) {
                final int id = p_id[0];

                final CharSequence msg;

                AppLog.e(TAG, "@@onError", error.getException());

                if (error.isConnectionError()) {
                    msg = TextUtils.expandTemplate(App.context.getString(R.string.version_download_network_error), title);
                } else {
                    msg = TextUtils.expandTemplate(App.context.getString(R.string.version_download_server_error), title);
                }

                App.context.startActivity(
                    AlertDialogActivity.createOkIntent(null, msg.toString())
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                );

                remove(id);

                App.getLbm().sendBroadcast(new Intent(VersionListFragment.ACTION_RELOAD));
            }
        });

        row.id = p_id[0];
        row.key = downloadKey;
        row.title = title;
        row.destPath = new File(downloadTempDirPath, downloadTempBasename).getAbsolutePath();
        row.attrs = new LinkedHashMap<>(attrs);
        currentByKey.put(downloadKey, row);
        currentById.put(row.id, row);
        displayNotifs();
    }

    void displayNotifs() {
        final Context context = App.context;

        final NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, context.getString(R.string.notification_channel_download_mapper_name), NotificationManager.IMPORTANCE_LOW);
        final NotificationManager nm = App.context.getSystemService(NotificationManager.class);
        if (nm != null) nm.createNotificationChannel(channel);

        final NotificationManagerCompat nmc = NotificationManagerCompat.from(context);
        for (final Row row : new ArrayList<>(currentById.values())) {
            final Notification n = new NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
                .setSmallIcon(android.R.drawable.stat_sys_download)
                .setContentTitle(row.title)
                .setContentText(row.totalBytes <= 0 ? "" : (Formatter.formatFileSize(context, row.currentBytes) + " / " + Formatter.formatFileSize(context, row.totalBytes)))
                .setShowWhen(false)
                .setProgress(1000, row.totalBytes == 0 ? 0 : (int) (1000 * row.currentBytes / row.totalBytes), false)
                .setOngoing(true)
                .build();

            nmc.notify(NOTIF_TAG, row.id, n);
        }
    }

    /**
     * Stop and remove from in-memory list and notification.
     */
    public void remove(final int id) {
        PRDownloader.cancel(id);

        final NotificationManagerCompat nmc = NotificationManagerCompat.from(App.context);
        nmc.cancel(NOTIF_TAG, id);

        final Row row = currentById.get(id);
        if (row != null) {
            currentByKey.remove(row.key);
            currentById.remove(row.id);
        }
    }
}

File: Alkitab/src/main/java/yuku/alkitab/base/util/DownloadMapper.kt
package yuku.alkitab.base.util

import android.app.DownloadManager
import android.content.Intent
import android.text.TextUtils
import androidx.annotation.VisibleForTesting
import androidx.work.Data
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkInfo
import androidx.work.WorkManager
import androidx.work.workDataOf
import java.io.File
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import yuku.alkitab.base.App
import yuku.alkitab.base.ac.AlertDialogActivity
import yuku.alkitab.base.br.VersionDownloadCompleteReceiver
import yuku.alkitab.base.events.AppEvents
import yuku.alkitab.debug.R

/**
 * In-process tracker for Bible version downloads, backed by WorkManager.
 *
 * Each in-flight download is represented by a [Row] keyed both by a stable string
 * [Row.key] (e.g. `"version:preset_name:TB"`) and by an internal int [Row.id] that
 * we allocate ourselves. The int id is what the rest of the app (notably
 * [VersionDownloadCompleteReceiver]) uses to identify a download; it preserves the
 * public surface of the previous PRDownloader-based implementation.
 *
 * Lifecycles:
 * - [enqueue] creates a [androidx.work.OneTimeWorkRequest] for [VersionDownloadWorker],
 *   registers the [Row], and launches an observer coroutine that mirrors WorkInfo
 *   state into [Row.dmStatus] (using Android `DownloadManager.STATUS_*` constants to
 *   keep the existing UI checks in `VersionListFragment` working unchanged).
 * - The observer also pumps progress data into the row and emits a throttled
 *   `versionListReload` so the list view re-renders.
 * - On terminal `SUCCEEDED`, the observer hands off to
 *   [VersionDownloadCompleteReceiver.onReceive], which finalises the file on a
 *   background thread and then calls [remove] in its `finally`.
 * - On terminal `FAILED`, the observer shows an error dialog and calls [remove].
 * - [remove] cancels both the observer coroutine and the underlying WorkManager
 *   work, then strips the row from the maps.
 */
class DownloadMapper private constructor() {

    private class Row(
        @JvmField val id: Int,
        @JvmField val key: String,
        @JvmField val title: String,
        @JvmField val destPath: String,
        @JvmField val workId: UUID,
        @JvmField val attrs: Map<String, String>,
    ) {
        @JvmField var currentBytes: Long = 0L
        @JvmField var totalBytes: Long = 0L
        @JvmField var dmStatus: Int = DownloadManager.STATUS_PENDING
        @JvmField var previouslyUpdatedProgressTime: Long = 0L
        @JvmField var observerJob: Job? = null
    }

    private val nextId = AtomicInteger(1)
    private val currentByKey = mutableMapOf<String, Row>()
    private val currentById = mutableMapOf<Int, Row>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    fun getStatus(downloadKey: String): Int = synchronized(this) {
        currentByKey[downloadKey]?.dmStatus ?: 0
    }

    fun getStatus(id: Int): Int = synchronized(this) {
        currentById[id]?.dmStatus ?: 0
    }

    fun getDownloadProgress(downloadKey: String): Float = synchronized(this) {
        val row = currentByKey[downloadKey] ?: return -1f
        if (row.totalBytes <= 0L) return -1f
        row.currentBytes.toFloat() / row.totalBytes.toFloat()
    }

    /** Must be called only after verifying that this id exists. */
    fun getAttrs(id: Int): Map<String, String>? = synchronized(this) {
        currentById[id]?.attrs
    }

    /** Must be called only after verifying that this id exists. */
    fun getDownloadedFilePath(id: Int): String? = synchronized(this) {
        currentById[id]?.destPath
    }

    fun enqueue(downloadKey: String, url: String, title: String, attrs: Map<String, String>) {
        val dir = downloadTempDirPath()
        val name = downloadTempBasename(downloadKey)
        val destPath = File(dir, name).absolutePath

        val request = OneTimeWorkRequest.Builder(VersionDownloadWorker::class.java)
            .setInputData(workDataOf(
                VersionDownloadWorker.KEY_URL to url,
                VersionDownloadWorker.KEY_DEST_PATH to destPath,
            ))
            .addTag(WORK_TAG)
            .build()

        val workManager = WorkManager.getInstance(App.context)
        workManager.enqueue(request)

        val id = nextId.getAndIncrement()
        val row = Row(
            id = id,
            key = downloadKey,
            title = title,
            destPath = destPath,
            workId = request.id,
            attrs = LinkedHashMap(attrs),
        )

        synchronized(this) {
            currentByKey[downloadKey] = row
            currentById[id] = row
        }

        // scope uses Dispatchers.Main and enqueue() is always invoked from the main
        // thread (UI click handlers), so this launch is guaranteed to run *after*
        // enqueue() returns — no race with the maps above.
        row.observerJob = scope.launch {
            workManager.getWorkInfoByIdFlow(request.id).collect { info ->
                if (info == null) {
                    silentlyDrop(row)
                    return@collect
                }
                when (info.state) {
                    WorkInfo.State.ENQUEUED -> row.dmStatus = DownloadManager.STATUS_PENDING

                    WorkInfo.State.RUNNING -> {
                        row.dmStatus = DownloadManager.STATUS_RUNNING
                        updateProgressFrom(row, info.progress)
                    }

                    WorkInfo.State.SUCCEEDED -> {
                        if (row.dmStatus != DownloadManager.STATUS_SUCCESSFUL) {
                            row.dmStatus = DownloadManager.STATUS_SUCCESSFUL
                            VersionDownloadCompleteReceiver.onReceive(row.id)
                        }
                    }

                    WorkInfo.State.FAILED -> {
                        if (row.dmStatus != DownloadManager.STATUS_FAILED) {
                            row.dmStatus = DownloadManager.STATUS_FAILED
                            handleFailure(row, info.outputData)
                        }
                    }

                    WorkInfo.State.CANCELLED -> silentlyDrop(row)

                    else -> Unit
                }
            }
        }

        AppEvents.emitVersionListReload()
    }

    private fun updateProgressFrom(row: Row, progressData: Data) {
        val current = progressData.getLong(VersionDownloadWorker.KEY_CURRENT_BYTES, -1L)
        val total = progressData.getLong(VersionDownloadWorker.KEY_TOTAL_BYTES, -1L)
        if (current < 0L) return
        row.currentBytes = current
        if (total >= 0L) row.totalBytes = total

        val now = System.currentTimeMillis()
        if (row.previouslyUpdatedProgressTime == 0L || now - row.previouslyUpdatedProgressTime > PROGRESS_NOTIFY_INTERVAL_MS) {
            row.previouslyUpdatedProgressTime = now
            AppEvents.emitVersionListReload()
        }
    }

    private fun handleFailure(row: Row, outputData: Data) {
        val errorType = outputData.getString(VersionDownloadWorker.KEY_ERROR_TYPE)
            ?: VersionDownloadWorker.ERROR_UNKNOWN
        val message = outputData.getString(VersionDownloadWorker.KEY_ERROR_MESSAGE) ?: errorType
        AppLog.e(TAG, "@@onError: downloadKey=${row.key} type=$errorType message=$message")

        val msg: CharSequence = when (errorType) {
            VersionDownloadWorker.ERROR_CONNECTION,
            VersionDownloadWorker.ERROR_CANCELLED ->
                TextUtils.expandTemplate(App.context.getString(R.string.version_download_network_error), row.title)

            VersionDownloadWorker.ERROR_STORAGE ->
                App.context.getString(R.string.version_download_saving_io_error)

            else ->
                TextUtils.expandTemplate(App.context.getString(R.string.version_download_server_error), row.title)
        }

        App.context.startActivity(
            AlertDialogActivity.createOkIntent(null, msg.toString())
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        )

        remove(row.id)
        AppEvents.emitVersionListReload()
    }

    private fun silentlyDrop(row: Row) {
        synchronized(this) {
            currentByKey.remove(row.key)
            currentById.remove(row.id)
        }
        row.observerJob?.cancel()
    }

    /** Stop and remove from in-memory list. */
    fun remove(id: Int) {
        val row: Row? = synchronized(this) {
            val r = currentById[id]
            if (r != null) {
                currentByKey.remove(r.key)
                currentById.remove(r.id)
            }
            r
        }
        if (row != null) {
            row.observerJob?.cancel()
            WorkManager.getInstance(App.context).cancelWorkById(row.workId)
        }
    }

    /**
     * Called by [VersionDownloadCompleteReceiver] once it has finished consuming
     * the downloaded temp file. Deletes the temp file and removes the in-memory
     * row.
     *
     * Why this exists: the temp file path is derived deterministically from the
     * download key, so leaving the file behind would make the next download with
     * the same key (typically a version *update*) see `existingBytes > 0` in
     * [VersionDownloadWorker] and try to resume from a stale offset against a
     * different file. The server then either replies 206 (corrupting the result
     * with a frankenstein of the old prefix and the new suffix) or 416 if the
     * new file is smaller, which surfaces to the user as a "Cannot connect to
     * server" generic error.
     *
     * Does not call `cancelWorkById`: the caller has already observed terminal
     * `SUCCEEDED` state, so the cancellation would be a no-op anyway.
     */
    fun consumeAndRemove(id: Int) {
        val row: Row? = synchronized(this) {
            val r = currentById[id]
            if (r != null) {
                currentByKey.remove(r.key)
                currentById.remove(r.id)
            }
            r
        }
        if (row != null) {
            row.observerJob?.cancel()
            @Suppress("ResultOfMethodCallIgnored")
            File(row.destPath).delete()
        }
    }

    /**
     * Test seam: inject a row directly without going through [enqueue] (which
     * would require WorkManager initialization and an actual HTTP fetch). Tests
     * use this to exercise the post-download lifecycle ([consumeAndRemove],
     * [remove]) against a controlled temp-file path.
     */
    @VisibleForTesting
    internal fun seedRowForTest(downloadKey: String, destPath: String): Int {
        val id = nextId.getAndIncrement()
        val row = Row(
            id = id,
            key = downloadKey,
            title = "test",
            destPath = destPath,
            workId = UUID.randomUUID(),
            attrs = emptyMap(),
        )
        synchronized(this) {
            currentByKey[downloadKey] = row
            currentById[id] = row
        }
        return id
    }

    private fun downloadTempDirPath(): String {
        val res = File(App.context.cacheDir, "DownloadMapper-tmp")
        res.mkdirs()
        return res.absolutePath
    }

    private fun downloadTempBasename(downloadKey: String): String {
        val safe = StringBuilder(downloadKey.length + 40)
        for (c in downloadKey) {
            if (c in 'A'..'Z' || c in 'a'..'z' || c in '0'..'9') safe.append(c)
        }
        return "DownloadMapper-$safe${downloadKey.hashCode()}.tmp"
    }

    companion object {
        private const val TAG = "DownloadMapper"
        private const val WORK_TAG = "VersionDownload"
        private const val PROGRESS_NOTIFY_INTERVAL_MS = 250L

        @JvmField
        val instance = DownloadMapper()
    }
}
